File: map/tool/tools/slam_export/cpp/src/slam_export_tool.cpp
#include "slam_export/slam_export_tool.hpp"

#include <algorithm>
#include <cmath>
#include <cstdint>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <map>
#include <set>
#include <sstream>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

#include <minizip/unzip.h>

namespace autodrive::slam_export {
namespace {

double round3(double value) {
    return std::round(value * 1000.0) / 1000.0;
}

std::string format3(double value) {
    std::ostringstream out;
    out << std::fixed << std::setprecision(3) << value;
    return out.str();
}

std::string base_name(const std::string& file_name) {
    const auto index = file_name.find_last_of('.');
    return index == std::string::npos ? file_name : file_name.substr(0, index);
}

std::string replace_extension(const std::string& file_name, const std::string& extension) {
    return base_name(file_name) + extension;
}

std::vector<unsigned char> read_zip_entry(unzFile archive, const char* name) {
    if (unzLocateFile(archive, name, 0) != UNZ_OK) {
        throw std::runtime_error(std::string("missing zip entry: ") + name);
    }
    unz_file_info64 info {};
    if (unzGetCurrentFileInfo64(archive, &info, nullptr, 0, nullptr, 0, nullptr, 0) != UNZ_OK) {
        throw std::runtime_error("failed to stat zip entry");
    }
    if (unzOpenCurrentFile(archive) != UNZ_OK) {
        throw std::runtime_error("failed to open zip entry");
    }
    std::vector<unsigned char> bytes(static_cast<std::size_t>(info.uncompressed_size));
    const auto read_size = unzReadCurrentFile(archive, bytes.data(), static_cast<unsigned int>(bytes.size()));
    unzCloseCurrentFile(archive);
    if (read_size < 0 || static_cast<std::size_t>(read_size) != bytes.size()) {
        throw std::runtime_error("failed to read zip entry");
    }
    return bytes;
}

PgmMetadata build_pgm(const nlohmann::json& manifest, const std::vector<SlamPoint>& points, double resolution, int padding_cells) {
    const auto browser = manifest.contains("browser_occupancy") && manifest["browser_occupancy"].is_object()
        ? manifest["browser_occupancy"]
        : nlohmann::json::object();
    const double occupancy_voxel = std::max(0.02, browser.value("voxel_size", resolution));
    const bool has_occupied_cells = browser.contains("occupied_cells") && browser["occupied_cells"].is_array() && !browser["occupied_cells"].empty();

    long min_cell_x = std::numeric_limits<long>::max();
    long max_cell_x = std::numeric_limits<long>::min();
    long min_cell_y = std::numeric_limits<long>::max();
    long max_cell_y = std::numeric_limits<long>::min();
    std::set<std::pair<long, long>> occupied;

    if (has_occupied_cells) {
        for (const auto& cell : browser["occupied_cells"]) {
            const long ix = std::lround(cell.value("ix", 0.0));
            const long iy = std::lround(cell.value("iy", 0.0));
            min_cell_x = std::min(min_cell_x, ix);
            max_cell_x = std::max(max_cell_x, ix);
            min_cell_y = std::min(min_cell_y, iy);
            max_cell_y = std::max(max_cell_y, iy);
            occupied.insert({ix, iy});
        }
    } else {
        if (points.empty()) {
            throw std::runtime_error("No radar points in SLAM");
        }
        for (const auto& point : points) {
            const long ix = std::lround(static_cast<double>(point.x) / resolution);
            const long iy = std::lround(static_cast<double>(point.y) / resolution);
            min_cell_x = std::min(min_cell_x, ix);
            max_cell_x = std::max(max_cell_x, ix);
            min_cell_y = std::min(min_cell_y, iy);
            max_cell_y = std::max(max_cell_y, iy);
            occupied.insert({ix, iy});
        }
    }

    const long padded_min_x = min_cell_x - padding_cells;
    const long padded_min_y = min_cell_y - padding_cells;
    const long padded_max_x = max_cell_x + padding_cells;
    const long padded_max_y = max_cell_y + padding_cells;
    const int width = std::max(1L, padded_max_x - padded_min_x + 1);
    const int height = std::max(1L, padded_max_y - padded_min_y + 1);
    std::vector<int> grid(static_cast<std::size_t>(width * height), 205);

    for (const auto& [ix_raw, iy_raw] : occupied) {
        const int ix = static_cast<int>(ix_raw - padded_min_x);
        const int iy = static_cast<int>(iy_raw - padded_min_y);
        const int flipped_y = height - 1 - iy;
        grid[static_cast<std::size_t>(flipped_y * width + ix)] = 0;
    }

    std::ostringstream pgm;
    pgm << "P2\n# Generated from SLAM occupancy\n" << width << " " << height << "\n255\n";
    for (int row = 0; row < height; ++row) {
        if (row > 0) {
            pgm << "\n";
        }
        const int start = row * width;
        for (int col = 0; col < width; ++col) {
            if (col > 0) {
                pgm << " ";
            }
            pgm << grid[static_cast<std::size_t>(start + col)];
        }
    }
    pgm << "\n";

    return PgmMetadata{
        pgm.str(),
        width,
        height,
        {round3(padded_min_x * occupancy_voxel), round3(padded_min_y * occupancy_voxel), 0.0},
        static_cast<int>(occupied.size()),
        {
            {"minX", round3(min_cell_x * occupancy_voxel)},
            {"maxX", round3(max_cell_x * occupancy_voxel)},
            {"minY", round3(min_cell_y * occupancy_voxel)},
            {"maxY", round3(max_cell_y * occupancy_voxel)},
        }
    };
}

std::string build_yaml(const std::string& file_name, double resolution, const std::array<double, 3>& origin) {
    std::ostringstream out;
    out << "image: " << replace_extension(file_name, ".pgm") << "\n";
    out << "mode: trinary\n";
    out << "resolution: " << format3(resolution) << "\n";
    out << "origin: [" << format3(origin[0]) << ", " << format3(origin[1]) << ", " << static_cast<int>(std::llround(origin[2])) << "]\n";
    out << "negate: 0\n";
    out << "occupied_thresh: 0.65\n";
    out << "free_thresh: 0.196";
    return out.str();
}

}  // namespace

LoadedSlam SlamExportTool::load(const std::string& path) {
    unzFile archive = unzOpen64(path.c_str());
    if (archive == nullptr) {
        throw std::runtime_error("failed to open .slam archive");
    }
    try {
        const auto manifest_bytes = read_zip_entry(archive, "manifest.json");
        const auto radar_bytes = read_zip_entry(archive, "radar_points.bin");
        unzClose(archive);

        const auto manifest_text = std::string(manifest_bytes.begin(), manifest_bytes.end());
        return LoadedSlam{nlohmann::json::parse(manifest_text), parse_radar_points(radar_bytes)};
    } catch (...) {
        unzClose(archive);
        throw;
    }
}

std::vector<SlamPoint> SlamExportTool::parse_radar_points(const std::vector<unsigned char>& bytes) {
    std::vector<SlamPoint> points;
    for (std::size_t offset = 0; offset + 12 <= bytes.size(); offset += 12) {
        union {
            std::uint32_t raw;
            float value;
        } x {}, y {}, intensity {};
        x.raw = static_cast<std::uint32_t>(bytes[offset])
            | (static_cast<std::uint32_t>(bytes[offset + 1]) << 8)
            | (static_cast<std::uint32_t>(bytes[offset + 2]) << 16)
            | (static_cast<std::uint32_t>(bytes[offset + 3]) << 24);
        y.raw = static_cast<std::uint32_t>(bytes[offset + 4])
            | (static_cast<std::uint32_t>(bytes[offset + 5]) << 8)
            | (static_cast<std::uint32_t>(bytes[offset + 6]) << 16)
            | (static_cast<std::uint32_t>(bytes[offset + 7]) << 24);
        intensity.raw = static_cast<std::uint32_t>(bytes[offset + 8])
            | (static_cast<std::uint32_t>(bytes[offset + 9]) << 8)
            | (static_cast<std::uint32_t>(bytes[offset + 10]) << 16)
            | (static_cast<std::uint32_t>(bytes[offset + 11]) << 24);
        points.push_back({x.value, y.value, intensity.value});
    }
    return points;
}

ExportArtifacts SlamExportTool::build_exports(
    const std::string& source_file,
    const nlohmann::json& manifest,
    const std::vector<SlamPoint>& points,
    double resolution,
    int padding_cells
) {
    const auto pgm = build_pgm(manifest, points, resolution, padding_cells);
    const auto yaml = build_yaml(source_file, resolution, pgm.origin);

    auto export_manifest = manifest;
    export_manifest.erase("browser_occupancy");
    export_manifest.erase("trajectory");

    nlohmann::json export_json = {
        {"source_file", source_file},
        {"map_yaml", {
            {"image", replace_extension(source_file, ".pgm")},
            {"mode", "trinary"},
            {"resolution", resolution},
            {"origin", {pgm.origin[0], pgm.origin[1], pgm.origin[2]}},
            {"negate", 0},
            {"occupied_thresh", 0.65},
            {"free_thresh", 0.196},
        }},
        {"pgm_meta", {
            {"width", pgm.width},
            {"height", pgm.height},
            {"occupied_cells", pgm.occupied_cells},
            {"bounds", pgm.bounds},
        }},
        {"manifest", export_manifest},
    };

    return ExportArtifacts{pgm, yaml, export_json.dump(2)};
}

ExportArtifacts SlamExportTool::export_to_dir(
    const std::string& slam_path,
    const std::string& output_dir,
    double resolution,
    int padding_cells
) {
    const auto loaded = load(slam_path);
    const auto source_file = std::filesystem::path(slam_path).filename().string();
    const auto artifacts = build_exports(source_file, loaded.manifest, loaded.radar_points, resolution, padding_cells);
    std::filesystem::create_directories(output_dir);
    const auto stem = base_name(source_file);
    std::ofstream(output_dir + "/" + stem + ".pgm") << artifacts.pgm.pgm_text;
    std::ofstream(output_dir + "/" + stem + ".yaml") << artifacts.yaml_text;
    std::ofstream(output_dir + "/" + stem + ".json") << artifacts.json_text;
    return artifacts;
}

}  // namespace autodrive::slam_export

File: tools/slam_export/java/src/com/autodrive/slamexport/SlamExportTool.java
package com.autodrive.slamexport;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import javax.script.ScriptException;

public final class SlamExportTool {
    private static final DecimalFormat THREE_DECIMAL = new DecimalFormat("0.000", DecimalFormatSymbols.getInstance(Locale.US));

    private SlamExportTool() {
    }

    public static LoadedSlam load(Path path) throws IOException {
        byte[] manifestBytes = null;
        byte[] radarBytes = null;
        try (ZipInputStream zis = new ZipInputStream(Files.newInputStream(path))) {
            ZipEntry entry;
            while ((entry = zis.getNextEntry()) != null) {
                if ("manifest.json".equals(entry.getName())) {
                    manifestBytes = readAllBytes(zis);
                } else if ("radar_points.bin".equals(entry.getName())) {
                    radarBytes = readAllBytes(zis);
                }
            }
        }
        if (manifestBytes == null || radarBytes == null) {
            throw new IOException(".slam archive must contain manifest.json and radar_points.bin");
        }
        return new LoadedSlam(parseManifest(new String(manifestBytes, StandardCharsets.UTF_8)), parseRadarPoints(radarBytes));
    }

    @SuppressWarnings("unchecked")
    public static Map<String, Object> parseManifest(String jsonText) throws IOException {
        ScriptEngine engine = new ScriptEngineManager().getEngineByName("javascript");
        if (engine == null) {
            throw new IOException("JavaScript engine unavailable for JSON parsing");
        }
        try {
            Object parsed = engine.eval("Java.asJSONCompatible(" + jsonText + ")");
            if (!(parsed instanceof Map)) {
                throw new IOException("manifest.json must decode to an object");
            }
            return (Map<String, Object>) deepCopy(parsed);
        } catch (ScriptException exc) {
            throw new IOException("Failed to parse manifest.json", exc);
        }
    }

    public static List<SlamPoint> parseRadarPoints(byte[] data) {
        List<SlamPoint> points = new ArrayList<SlamPoint>();
        ByteBuffer buffer = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN);
        while (buffer.remaining() >= 12) {
            points.add(new SlamPoint(buffer.getFloat(), buffer.getFloat(), buffer.getFloat()));
        }
        return points;
    }

    public static ExportArtifacts buildExports(String sourceFile, Map<String, Object> manifest, List<SlamPoint> points, double resolution, int paddingCells) {
        PgmMetadata pgm = buildPgm(sourceFile, manifest, points, resolution, paddingCells);
        String yaml = buildYamlText(sourceFile, resolution, pgm.getOrigin());
        Map<String, Object> exportManifest = deepCopyMap(manifest);
        exportManifest.remove("browser_occupancy");
        exportManifest.remove("trajectory");

        Map<String, Object> mapYaml = new LinkedHashMap<String, Object>();
        mapYaml.put("image", replaceExtension(sourceFile, ".pgm"));
        mapYaml.put("mode", "trinary");
        mapYaml.put("resolution", resolution);
        mapYaml.put("origin", pgm.getOrigin());
        mapYaml.put("negate", 0);
        mapYaml.put("occupied_thresh", 0.65);
        mapYaml.put("free_thresh", 0.196);

        Map<String, Object> pgmMeta = new LinkedHashMap<String, Object>();
        pgmMeta.put("width", pgm.getWidth());
        pgmMeta.put("height", pgm.getHeight());
        pgmMeta.put("occupied_cells", pgm.getOccupiedCells());
        pgmMeta.put("bounds", pgm.getBounds());

        Map<String, Object> exportJson = new LinkedHashMap<String, Object>();
        exportJson.put("source_file", sourceFile);
        exportJson.put("map_yaml", mapYaml);
        exportJson.put("pgm_meta", pgmMeta);
        exportJson.put("manifest", exportManifest);

        return new ExportArtifacts(pgm, yaml, toJson(exportJson, 0));
    }

    public static ExportArtifacts export(Path slamPath, Path outputDir, double resolution, int paddingCells) throws IOException {
        LoadedSlam loaded = load(slamPath);
        ExportArtifacts artifacts = buildExports(slamPath.getFileName().toString(), loaded.getManifest(), loaded.getRadarPoints(), resolution, paddingCells);
        Files.createDirectories(outputDir);
        String stem = baseName(slamPath.getFileName().toString());
        Files.writeString(outputDir.resolve(stem + ".pgm"), artifacts.getPgmText(), StandardCharsets.UTF_8);
        Files.writeString(outputDir.resolve(stem + ".yaml"), artifacts.getYamlText(), StandardCharsets.UTF_8);
        Files.writeString(outputDir.resolve(stem + ".json"), artifacts.getJsonText(), StandardCharsets.UTF_8);
        return artifacts;
    }

    @SuppressWarnings("unchecked")
    private static PgmMetadata buildPgm(String sourceFile, Map<String, Object> manifest, List<SlamPoint> points, double resolution, int paddingCells) {
        Map<String, Object> browser = manifest.get("browser_occupancy") instanceof Map
            ? (Map<String, Object>) manifest.get("browser_occupancy")
            : null;
        double occupancyVoxel = Math.max(0.02, getDouble(browser, "voxel_size", resolution));
        List<Object> occupiedCells = browser != null && browser.get("occupied_cells") instanceof List
            ? (List<Object>) browser.get("occupied_cells")
            : null;

        long minCellX = Long.MAX_VALUE;
        long maxCellX = Long.MIN_VALUE;
        long minCellY = Long.MAX_VALUE;
        long maxCellY = Long.MIN_VALUE;
        Set<String> occupied = new LinkedHashSet<String>();

        if (occupiedCells != null && !occupiedCells.isEmpty()) {
            for (Object item : occupiedCells) {
                Map<String, Object> cell = (Map<String, Object>) item;
                long ix = Math.round(getDouble(cell, "ix", 0.0));
                long iy = Math.round(getDouble(cell, "iy", 0.0));
                minCellX = Math.min(minCellX, ix);
                maxCellX = Math.max(maxCellX, ix);
                minCellY = Math.min(minCellY, iy);
                maxCellY = Math.max(maxCellY, iy);
                occupied.add(ix + ":" + iy);
            }
        } else {
            if (points.isEmpty()) {
                throw new IllegalArgumentException("No radar points in SLAM");
            }
            for (SlamPoint point : points) {
                long ix = Math.round(point.getX() / resolution);
                long iy = Math.round(point.getY() / resolution);
                minCellX = Math.min(minCellX, ix);
                maxCellX = Math.max(maxCellX, ix);
                minCellY = Math.min(minCellY, iy);
                maxCellY = Math.max(maxCellY, iy);
                occupied.add(ix + ":" + iy);
            }
        }

        long paddedMinX = minCellX - paddingCells;
        long paddedMinY = minCellY - paddingCells;
        long paddedMaxX = maxCellX + paddingCells;
        long paddedMaxY = maxCellY + paddingCells;
        int width = Math.max(1, (int) (paddedMaxX - paddedMinX + 1));
        int height = Math.max(1, (int) (paddedMaxY - paddedMinY + 1));
        int[] grid = new int[width * height];
        for (int i = 0; i < grid.length; i++) {
            grid[i] = 205;
        }

        for (String cellKey : occupied) {
            String[] parts = cellKey.split(":");
            int ix = (int) (Long.parseLong(parts[0]) - paddedMinX);
            int iy = (int) (Long.parseLong(parts[1]) - paddedMinY);
            int flippedY = height - 1 - iy;
            grid[flippedY * width + ix] = 0;
        }

        StringBuilder rows = new StringBuilder();
        rows.append("P2\n# Generated from SLAM occupancy\n");
        rows.append(width).append(" ").append(height).append("\n255\n");
        for (int row = 0; row < height; row++) {
            if (row > 0) {
                rows.append("\n");
            }
            int start = row * width;
            for (int col = 0; col < width; col++) {
                if (col > 0) {
                    rows.append(" ");
                }
                rows.append(grid[start + col]);
            }
        }
        rows.append("\n");

        List<Double> origin = List.of(round3(paddedMinX * occupancyVoxel), round3(paddedMinY * occupancyVoxel), 0.0);
        Map<String, Double> bounds = new LinkedHashMap<String, Double>();
        bounds.put("minX", round3(minCellX * occupancyVoxel));
        bounds.put("maxX", round3(maxCellX * occupancyVoxel));
        bounds.put("minY", round3(minCellY * occupancyVoxel));
        bounds.put("maxY", round3(maxCellY * occupancyVoxel));
        return new PgmMetadata(rows.toString(), width, height, origin, occupied.size(), bounds);
    }

    private static String buildYamlText(String fileName, double resolution, List<Double> origin) {
        return "image: " + replaceExtension(fileName, ".pgm") + "\n"
            + "mode: trinary\n"
            + "resolution: " + THREE_DECIMAL.format(resolution) + "\n"
            + "origin: [" + THREE_DECIMAL.format(origin.get(0)) + ", " + THREE_DECIMAL.format(origin.get(1)) + ", " + Math.round(origin.get(2)) + "]\n"
            + "negate: 0\n"
            + "occupied_thresh: 0.65\n"
            + "free_thresh: 0.196";
    }

    private static byte[] readAllBytes(ZipInputStream zis) throws IOException {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        byte[] buffer = new byte[4096];
        int read;
        while ((read = zis.read(buffer)) >= 0) {
            out.write(buffer, 0, read);
        }
        return out.toByteArray();
    }

    private static String replaceExtension(String fileName, String extension) {
        return baseName(fileName) + extension;
    }

    private static String baseName(String fileName) {
        int index = fileName.lastIndexOf('.');
        return index >= 0 ? fileName.substring(0, index) : fileName;
    }

    private static double getDouble(Map<String, Object> map, String key, double defaultValue) {
        if (map == null) {
            return defaultValue;
        }
        Object value = map.get(key);
        if (value instanceof Number) {
            return ((Number) value).doubleValue();
        }
        if (value instanceof String) {
            try {
                return Double.parseDouble((String) value);
            } catch (NumberFormatException ignored) {
                return defaultValue;
            }
        }
        return defaultValue;
    }

    @SuppressWarnings("unchecked")
    private static Object deepCopy(Object value) {
        if (value instanceof Map) {
            Map<String, Object> copy = new LinkedHashMap<String, Object>();
            for (Map.Entry<String, Object> entry : ((Map<String, Object>) value).entrySet()) {
                copy.put(entry.getKey(), deepCopy(entry.getValue()));
            }
            return copy;
        }
        if (value instanceof List) {
            List<Object> copy = new ArrayList<Object>();
            for (Object item : (List<Object>) value) {
                copy.add(deepCopy(item));
            }
            return copy;
        }
        return value;
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> deepCopyMap(Map<String, Object> value) {
        return (Map<String, Object>) deepCopy(value);
    }

    @SuppressWarnings("unchecked")
    private static String toJson(Object value, int indent) {
        if (value == null) {
            return "null";
        }
        if (value instanceof String) {
            return "\"" + escapeJson((String) value) + "\"";
        }
        if (value instanceof Number || value instanceof Boolean) {
            return String.valueOf(value);
        }
        if (value instanceof Map) {
            StringBuilder builder = new StringBuilder();
            builder.append("{");
            boolean first = true;
            for (Map.Entry<String, Object> entry : ((Map<String, Object>) value).entrySet()) {
                if (!first) {
                    builder.append(",");
                }
                builder.append("\n").append(spaces(indent + 2));
                builder.append("\"").append(escapeJson(entry.getKey())).append("\": ");
                builder.append(toJson(entry.getValue(), indent + 2));
                first = false;
            }
            if (!first) {
                builder.append("\n").append(spaces(indent));
            }
            builder.append("}");
            return builder.toString();
        }
        if (value instanceof List) {
            List<Object> list = (List<Object>) value;
            StringBuilder builder = new StringBuilder();
            builder.append("[");
            for (int i = 0; i < list.size(); i++) {
                if (i > 0) {
                    builder.append(", ");
                }
                builder.append(toJson(list.get(i), indent));
            }
            builder.append("]");
            return builder.toString();
        }
        return "\"" + escapeJson(String.valueOf(value)) + "\"";
    }

    private static String spaces(int count) {
        return " ".repeat(Math.max(0, count));
    }

    private static String escapeJson(String text) {
        return text
            .replace("\\", "\\\\")
            .replace("\"", "\\\"")
            .replace("\n", "\\n")
            .replace("\r", "\\r")
            .replace("\t", "\\t");
    }

    private static double round3(double value) {
        return Double.parseDouble(THREE_DECIMAL.format(value));
    }
}
